src/drivers/riscv_plic0.c

```diff
@@ -30,36 +30,45 @@
 extern metal_interrupt_handler_t __metal_global_interrupt_table[];
 
 static __inline__ unsigned int __metal_plic0_claim_interrupt(uint32_t hartid) {
-    return PLIC_REGW(PLIC_CONTEXT_BASE(hartid) +
-                     METAL_RISCV_PLIC0_CONTEXT_CLAIM);
+    return PLIC_REGW(
+        METAL_RISCV_PLIC0_CONTEXT_BASE +
+        (PLIC_CONTEXT_ID(hartid) * METAL_RISCV_PLIC0_PER_CONTEXT_OFFSET) +
+        METAL_RISCV_PLIC0_CONTEXT_CLAIM);
 }
 
 static __inline__ void __metal_plic0_complete_interrupt(int hartid,
                                                         unsigned int id) {
-    PLIC_REGW(PLIC_CONTEXT_BASE(hartid) + METAL_RISCV_PLIC0_CONTEXT_CLAIM) = id;
+    PLIC_REGW(METAL_RISCV_PLIC0_CONTEXT_BASE +
+              (PLIC_CONTEXT_ID(hartid) * METAL_RISCV_PLIC0_PER_CONTEXT_OFFSET) +
+              METAL_RISCV_PLIC0_CONTEXT_CLAIM) = id;
 }
 
 static __inline__ int
 __metal_riscv_plic0_set_threshold(int hartid, unsigned int threshold) {
-    PLIC_REGW(PLIC_CONTEXT_BASE(hartid) + METAL_RISCV_PLIC0_CONTEXT_THRESHOLD) =
-        threshold;
+    PLIC_REGW(METAL_RISCV_PLIC0_CONTEXT_BASE +
+              (PLIC_CONTEXT_ID(hartid) * METAL_RISCV_PLIC0_PER_CONTEXT_OFFSET) +
+              METAL_RISCV_PLIC0_CONTEXT_CLAIM) = threshold;
     return 0;
 }
 
 static __inline__ unsigned int __metal_riscv_plic0_get_threshold(int hartid) {
-    return PLIC_REGW(PLIC_CONTEXT_BASE(hartid) +
-                     METAL_RISCV_PLIC0_CONTEXT_THRESHOLD);
+    return PLIC_REGW(
+        METAL_RISCV_PLIC0_CONTEXT_BASE +
+        (PLIC_CONTEXT_ID(hartid) * METAL_RISCV_PLIC0_PER_CONTEXT_OFFSET) +
+        METAL_RISCV_PLIC0_CONTEXT_CLAIM);
 }
 
 static __inline__ int __metal_plic0_enable(int hartid, int id) {
     if (id >= METAL_RISCV_PLIC0_0_RISCV_NDEV) {
         return -1;
     }
 
+    const uintptr_t id_offset = (id / 32) * 4;
+    const uint32_t enable_mask = (1 << id % 32);
```

Comment:
parens here to avoid confusion over precedence (yeah, I know it's correct as-is, still made me pause :-)